File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.js
function solution(weights) {
  let answer = 0;
  const map = new Map();
  const ratio = [1, 3 / 2, 2, 4 / 3];

  weights
    .sort((a, b) => b - a)
    .forEach((weight) => {
      ratio.forEach((r) => {
        if (map.has(weight * r)) answer += map.get(weight * r);
      });
      map.set(weight, (map.get(weight) || 0) + 1);
    });

  return answer;
}

File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.java
import java.util.*;

class Solution {
    public long solution(int[] weights) {
        long answer = 0;
        HashMap<Double, Integer> map = new HashMap<>();
        Arrays.sort(weights);
        double[] ratio = new double[]{1.0, 1.5, 2.0, 4.0 / 3.0};

        for (Integer weight : weights) {
            double w = Double.valueOf(weight);

            if (map.containsKey(w)) {
                answer += map.get(w);
            }

            for (double v : ratio) {
                map.put(w * v, map.getOrDefault(w * v, 0) + 1);
            }
        }

        return answer;
    }
}